optuna_integration/_lightgbm_tuner/optimize.py

```diff
@@ -59,6 +59,14 @@
 _logger = optuna.logging.get_logger(__name__)
 
 
+def _get_custom_objective(lgbm_kwargs: dict[str, Any]) -> Callable[..., Any] | None:
```

Comment:
Please use `Protocol` instead. To do so, you need to add an import statement  `from typing import Protocol`

```suggestion
class _CustomObjectiveType(Protocol):
    def __call__(self, preds: np.ndarray, train: "lgb.Dataset") -> tuple[np.ndarray, np.ndarray]:
        # NOTE(nabenabe): Please check the official documentation:
        # https://lightgbm.readthedocs.io/en/stable/pythonapi/lightgbm.train.html
        # return grad, hess
        raise NotImplementedError


def _get_custom_objective(lgbm_kwargs: dict[str, Any]) -> _CustomObjectiveType | None:
```